nucliadb/nucliadb/common/cluster/manager.py

```diff
@@ -437,31 +461,32 @@ async def add_resource(
 
 
 def choose_node(
-    shard: writer_pb2.ShardObject, target_replicas: Optional[list[str]] = None
+    shard: writer_pb2.ShardObject, read_only: bool = False
 ) -> tuple[AbstractIndexNode, str, str]:
     """
-    Choose an arbitrary node storing `shard`. If passed, attempt to choose only between
-    nodes containing any of `target_replicas`.
+    Choose an arbitrary node storing `shard`.
     """
-    target_replicas = target_replicas or []
-
-    shuffled_replicas = [(x.shard.id, x.node) for x in shard.replicas]
-    random.shuffle(shuffled_replicas)
-
-    head_nodes = []
-    tail_nodes = []
-    for replica_id, node_id in shuffled_replicas:
-        if replica_id in target_replicas:
-            head_nodes.append((replica_id, node_id))
-        else:
-            tail_nodes.append((replica_id, node_id))
-
-    for replica_id, node_id in head_nodes + tail_nodes:
+    potential_nodes = []
+    for shardreplica in shard.replicas:
+        node_id = shardreplica.node
+        replica_id = shardreplica.shard.id
         node = get_index_node(node_id)
         if node is not None:
-            return node, replica_id, node_id
+            potential_nodes.append((replica_id, node))
+
+        if read_only:
+            for read_replica_node_id in get_read_replica_node_ids(node_id):
+                read_replica_node = get_index_node(read_replica_node_id)
+                if read_replica_node is None:
+                    continue
+                potential_nodes.append((replica_id, read_replica_node))
+
+    if len(potential_nodes) == 0:
+        raise NoHealthyNodeAvailable("Could not find a node to query")
 
-    raise KeyError("Could not find a node to query")
+    selected_node: AbstractIndexNode
+    replica_id, selected_node = random.choice(potential_nodes)
```

Comment:
If we go for a random choice approach here, we could be giving inconsistent results during  the time that the changes after a write are being replicated to the read replicas.

By inconsistent I mean that some requests would hit the primary shard and return the "up to date" results and others would hit the replicas and return the unsynced results.

Maybe we could cache the `choose_node` decision for some time?
Just throwing a thought here, I am not sure if this is a big deal...

Reply:
Another thought:
Since we know that search performance can be impacted by merges / gc, would it make sense to give priority to read only replicas here?

This way we make sure we query on the right segments.

Reply:
IMO, for the system we have an how writes are coming in, I think this is a good initial approach.

I agree with every one of your points; however, I think there are other things we can work toward before we add more up front decisions in this component.

For example:
1. we should reduce the time of when a change to an index happens to when it is replicated. Right now, we are really naive in our first approach. But let's measure and focus on bringing this time down long term instead of caching
2. weighting the node selection -- I really agree! But IMO, let's do something very simple to start with
3. finally, remember, we'd like to build an index coordinator. This component implement is somewhat temporary hopefully but maybe I'm overly hopeful